sys/net/gnrc/link_layer/lwmac/lwmac.c

```diff
@@ -0,0 +1,745 @@
+        return;
+
+    if(newstate >= STATE_COUNT) {
+        LOG_ERROR("Trying to set invalid state %u\n", newstate);
+        return;
+    }
+
+    /* Already change state, but might be reverted to oldstate when needed */
+    lwmac.state = newstate;
+
+    /* Actions when leaving old state */
+    switch(oldstate)
+    {
+    case RECEIVING:
+    case TRANSMITTING:
+        /* Enable duty cycling again */
+        rtt_handler(LWMAC_EVENT_RTT_RESUME);
+        break;
+
+    case SLEEPING:
+        lwmac_clear_timeout(&lwmac, TIMEOUT_WAKEUP_PERIOD);
+        break;
+
+    default:
+        break;
+    }
+
+    /* Actions when entering new state */
+    switch(newstate)
+    {
+    /*********************** Operation states *********************************/
+    case LISTENING:
+        _set_netdev_state(&lwmac, NETOPT_STATE_IDLE);
+        break;
+
+    case SLEEPING:
+        /* Put transceiver to sleep */
+        _set_netdev_state(&lwmac, NETOPT_STATE_SLEEP);
+        /* We may have come here through RTT handler, so timeout may still be active */
+        lwmac_clear_timeout(&lwmac, TIMEOUT_WAKEUP_PERIOD);
+        /* Return immediately, so no rescheduling */
+        return;
+
+    /* Trying to send data */
+    case TRANSMITTING:
+        rtt_handler(LWMAC_EVENT_RTT_PAUSE); /* No duty cycling while RXing */
+        _set_netdev_state(&lwmac, NETOPT_STATE_IDLE);  /* Power up netdev */
+        break;
+
+    /* Receiving incoming data */
+    case RECEIVING:
+        rtt_handler(LWMAC_EVENT_RTT_PAUSE); /* No duty cycling while TXing */
+        _set_netdev_state(&lwmac, NETOPT_STATE_IDLE);  /* Power up netdev */
+        break;
+
+    case STOPPED:
+        _set_netdev_state(&lwmac, NETOPT_STATE_OFF);
+        break;
+
+    /*********************** Control states ***********************************/
+    case START:
+        rtt_handler(LWMAC_EVENT_RTT_START);
+        lwmac_set_state(LISTENING);
+        break;
+
+    case STOP:
+        rtt_handler(LWMAC_EVENT_RTT_STOP);
+        lwmac_set_state(STOPPED);
+        break;
+
+    case RESET:
+        LOG_WARNING("Reset not yet implemented\n");
+        lwmac_set_state(STOP);
+        lwmac_set_state(START);
+        break;
+
+    /**************************************************************************/
+    default:
+        LOG_DEBUG("No actions for entering state %u\n", newstate);
+        return;
+    }
+
+    lwmac_schedule_update();
+}
+
+/******************************************************************************/
+
+/* Main state machine. Call whenever something happens */
+// TODO: Don't use global variables
+bool lwmac_update(void)
+{
+    lwmac.needs_rescheduling = false;
+
+    switch(lwmac.state)
+    {
+    case SLEEPING:
+
+        /* If a packet is scheduled, no other (possible earlier) packet can be
+         * sent before the first one is handled, even no broadcast
+         */
+        if( !lwmac_timeout_is_running(&lwmac, TIMEOUT_WAIT_FOR_DEST_WAKEUP) ) {
```

Comment:
Sorry for the (really) delayed response here.

> When the device is in the sleeping period, it will be in the standby mode (low-power mode) right?

Yes and no. There's a difference of the cpu state and the LwMAC state. If the cpu is sleeping, nothing happens obviously. Now imagine the case there's thread A (which is your application), that sends packets over the network is not synchronized to any MAC layer periods. While the MAC layer is "sleeping", thread A can initiate sending a packet which then arrives at the MAC layer. While the radio is in it's sleeping period, the MAC layer will still handle the incoming packet (and then finally send it at some time).

The particular case here is the following:
* MAC is sleeping
* thread A sends a packet to MAC layer
* MAC queues incoming packet
* MAC checks if there are packets in TX queue (yes!) and then starts timeout for neighbors wakeup phase
* MAC waits until timeout expires and then goes to TRANSMITTING state

While the timeout is running, a wake lock is requested, because, as you identified correctly, timeouts currently are based on xtimer and therefore only work when the cpu is running.

~~~c
/* Stop dutycycling, we're preparing to send. This prevents the
 * timeout arriving late, so that the destination phase would
 * be missed. */
// TODO: bad for power savings
rtt_handler(LWMAC_EVENT_RTT_PAUSE);
~~~